Include/cpython/unicodeobject.h

```diff
@@ -287,160 +339,172 @@ PyAPI_FUNC(int) _PyUnicode_CheckConsistency(
+    }
+    else {
+        return _PyUnicode_NONCOMPACT_DATA(op);
+    }
+}
+#define PyUnicode_DATA(op) PyUnicode_DATA(_PyObject_CAST(op))
+
 /* Return pointers to the canonical representation cast to unsigned char,
    Py_UCS2, or Py_UCS4 for direct character access.
    No checks are performed, use PyUnicode_KIND() before to ensure
    these will work correctly. */
-
 #define PyUnicode_1BYTE_DATA(op) ((Py_UCS1*)PyUnicode_DATA(op))
 #define PyUnicode_2BYTE_DATA(op) ((Py_UCS2*)PyUnicode_DATA(op))
 #define PyUnicode_4BYTE_DATA(op) ((Py_UCS4*)PyUnicode_DATA(op))
 
-/* Return one of the PyUnicode_*_KIND values defined above. */
-#define PyUnicode_KIND(op) \
-    (assert(PyUnicode_IS_READY(op)),            \
-     _PyASCIIObject_CAST(op)->state.kind)
-
-/* Return a void pointer to the raw unicode buffer. */
-#define _PyUnicode_COMPACT_DATA(op)                     \
-    (PyUnicode_IS_ASCII(op) ?                           \
-     ((void*)(_PyASCIIObject_CAST(op) + 1)) :           \
-     ((void*)(_PyCompactUnicodeObject_CAST(op) + 1)))
-
-#define _PyUnicode_NONCOMPACT_DATA(op)                  \
-    (assert(_PyUnicodeObject_CAST(op)->data.any),       \
-     (_PyUnicodeObject_CAST(op)->data.any))
-
-#define PyUnicode_DATA(op) \
-    (PyUnicode_IS_COMPACT(op) ? _PyUnicode_COMPACT_DATA(op) :   \
-     _PyUnicode_NONCOMPACT_DATA(op))
-
-/* In the access macros below, "kind" may be evaluated more than once.
-   All other macro parameters are evaluated exactly once, so it is safe
-   to put side effects into them (such as increasing the index). */
+/* Returns the length of the unicode string. The caller has to make sure that
+   the string has it's canonical representation set before calling
+   this function.  Call PyUnicode_(FAST_)Ready to ensure that. */
+static inline Py_ssize_t PyUnicode_GET_LENGTH(PyObject *op) {
+    assert(PyUnicode_IS_READY(op));
+    return _PyASCIIObject_CAST(op)->length;
+}
+#define PyUnicode_GET_LENGTH(op) PyUnicode_GET_LENGTH(_PyObject_CAST(op))
 
-/* Write into the canonical representation, this macro does not do any sanity
+/* Write into the canonical representation, this function does not do any sanity
    checks and is intended for usage in loops.  The caller should cache the
-   kind and data pointers obtained from other macro calls.
+   kind and data pointers obtained from other function calls.
    index is the index in the string (starts at 0) and value is the new
    code point value which should be written to that location. */
+static inline void PyUnicode_WRITE(unsigned int kind, void *data,
+                                   Py_ssize_t index, Py_UCS4 value)
+{
+    switch (kind) {
+    case PyUnicode_1BYTE_KIND:
+        ((Py_UCS1 *)data)[index] = (Py_UCS1)value;
+        break;
+    case PyUnicode_2BYTE_KIND:
+        ((Py_UCS2 *)data)[index] = (Py_UCS2)value;
+        break;
+    default:
+        assert(kind == PyUnicode_4BYTE_KIND);
+        ((Py_UCS4 *)data)[index] = value;
+    }
```

Comment:
It seems that the switch be collapsed into a single memcpy without losing readability:

```suggestion
    const size_t width = kind;
    assert(width == 1 || width == 2 || width == 4);
    memcpy(data + index * width, value, width);
```